crates/glue/writing_a_glue_plugin.md

```diff
@@ -0,0 +1,441 @@
+  - [Unit](#unit)
+  - [Bool](#bool)
+  - [Num](#num)
+  - [RocStr, RocList, RocDict, RocSet, RocBox](#rocstr-roclist-rocdict-rocset-rocbox)
+  - [Struct](#struct)
+    - [HasNoClosure](#hasnoclosure)
+    - [HasClosure](#hasclosure)
+  - [TagUnion](#tagunion)
+    - [NonRecursive](#nonrecursive)
+    - [Enumeration](#enumeration)
+    - [SingleTagStruct](#singletagstruct)
+    - [Recursive](#recursive)
+      - [With pointer tagging](#with-pointer-tagging)
+      - [Without pointer tagging](#without-pointer-tagging)
+    - [NullableWrapped](#nullablewrapped)
+    - [NullableUnwrapped](#nullableunwrapped)
+    - [NonNullableUnwrapped](#nonnullableunwrapped)
+  - [EmptyTagUnion](#emptytagunion)
+  - [TagUnionPayload](#tagunionpayload)
+  - [Function](#function)
+  - [Unsized](#unsized)
+  - [RecursivePointer](#recursivepointer)
+
+# The Types data structure
+
+`Types` is a data structure that stores the layout information of all types on the host-application FFI boundry for a specific architecture. This includes these types' size, alignment, order of declaration, and their `Shape`.
+
+# The Shape data structure
+
+`Shape` is where `roc glue` stores all layout information of a type specific to its kind. It consists of a tag union with the following tags:
+
+## Unit
+
+An empty record `{}`, Roc's anonymous zero-sized type.
+
+## Bool
+
+A one-byte type representing `true` as 1 and `false` as 0.
+
+## Num
+
+A number type. Either a signed integer (I8 through I128), an unsigned integer (U8 through U128), an IEEE 754 floating point number (F32 or F64), or a Roc-specific fixed-point number type ([Dec](https://www.roc-lang.org/builtins/Num#Dec)).
+
+## RocStr, RocList, RocDict, RocSet, RocBox
+
+Roc builtin datatypes. The Roc compiler is planned to expose parts of the standard library to glue code via C FFI to avoid having the reimplement them by hand in every host language. In the meantime, Rust implementations of the `Str`, `List`, and `Box` datatypes can be found under `crates/roc_std` for reference.
+
+## Struct
+
+A Roc record type. Also known as a struct or a product type. Its `fields` field can be one of two variants:
+
+### HasNoClosure
+
+Contains a list of fields with a name and type. Fields are arranged in memory in listed order as they would be in a C struct. Fields are not guaranteed to be in order of declaration in Roc.
+
+```roc
+# in `platform.roc`
+Color : { red : U8, green : U8, blue : U8 }
+```
+```roc
+# shape reported by `roc glue`
+Struct {
+    name: "Color",
+    fields: (HasNoClosure [
+        # `@TypeId 1` == `Num U8`
+        {id: (@TypeId 1), name: "blue"},
+        {id: (@TypeId 1), name: "green"},
+        {id: (@TypeId 1), name: "red"},
+    ]),
+}
+```
+```c
+// example C layout
+struct Color {
+    uint8_t blue;
+    uint8_t green;
+    uint8_t red;
+};
+```
+
+### HasClosure
+
+(TODO: no glue spec currently supports this. Unsure if this either does not work or if nobody who knows how has gotten around to it.)
```

Comment:
I have done it manually, but it really is not fun. Probably want boxed closures. Then it should work basically the same as no closure case. Just some of the fields will be boxes containing closure data.